File: backend/app/schemas/logistics.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class InventoryItem(BaseModel):
    id: int
    product_name: str
    batch_code: str
    origin: str
    state: str
    quantity: int
    volume_m3: float
    warehouse_location: str
    entry_date: str
    status: str


class MovementBase(BaseModel):
    folio: str = Field(min_length=4)
    movement_type: Literal["entrada", "salida"]
    partner: str
    quantity: int = Field(gt=0)
    volume_m3: float = Field(gt=0)
    state: str
    travel_time_hours: float = Field(ge=0)
    responsible: str
    movement_date: str
    batch_code: str
    notes: str = ""


class MovementCreate(MovementBase):
    product_name: str
    origin: str
    warehouse_location: str
    status: str


class Movement(MovementBase):
    id: int


class LogisticsTemplateRow(BaseModel):
    movement_date: str
    folio: str
    provider_or_destination: str
    quantity: int
    volume_m3: float
    state: str
    travel_time_hours: float
    responsible: str
    movement_type: Literal["entrada", "salida"]
    batch_code: str


class RouteBase(BaseModel):
    state_name: str
    destination: str
    travel_time_hours: float = Field(ge=0)
    status: str
    responsible: str
    capacity_units: int = Field(gt=0)
    last_maintenance: str


class RouteCreate(RouteBase):
    pass


class Route(RouteBase):
    id: int


class DistributionStateSummary(BaseModel):
    state_name: str
    route_count: int
    active_routes: int
    maintenance_routes: int
    total_capacity_units: int
    average_travel_time_hours: float


class DashboardMetrics(BaseModel):
    total_units: int
    total_volume_m3: float
    active_routes: int
    monthly_movements: int
    estimated_waste_reduction_pct: float

File: frontend/src/types/logistics.ts
export type DashboardMetrics = {
  total_units: number;
  total_volume_m3: number;
  active_routes: number;
  monthly_movements: number;
  estimated_waste_reduction_pct: number;
};

export type InventoryItem = {
  id: number;
  product_name: string;
  batch_code: string;
  origin: string;
  state: string;
  quantity: number;
  volume_m3: number;
  warehouse_location: string;
  entry_date: string;
  status: string;
};

export type Movement = {
  id: number;
  folio: string;
  movement_type: 'entrada' | 'salida';
  partner: string;
  quantity: number;
  volume_m3: number;
  state: string;
  travel_time_hours: number;
  responsible: string;
  movement_date: string;
  batch_code: string;
  notes: string;
};

export type MovementPayload = Omit<Movement, 'id'> & {
  product_name: string;
  origin: string;
  warehouse_location: string;
  status: string;
};

export type LogisticsTemplateRow = {
  movement_date: string;
  folio: string;
  provider_or_destination: string;
  quantity: number;
  volume_m3: number;
  state: string;
  travel_time_hours: number;
  responsible: string;
  movement_type: 'entrada' | 'salida';
  batch_code: string;
};

export type Route = {
  id: number;
  state_name: string;
  destination: string;
  travel_time_hours: number;
  status: string;
  responsible: string;
  capacity_units: number;
  last_maintenance: string;
};

export type RoutePayload = Omit<Route, 'id'>;

export type DistributionStateSummary = {
  state_name: string;
  route_count: number;
  active_routes: number;
  maintenance_routes: number;
  total_capacity_units: number;
  average_travel_time_hours: number;
};
